### ISSUE TITLE
hyperspectra with multiple files

### ISSUE THREAD

**Issue by GioB81** [Quasars/orange-spectroscopy#562]
Hi,

when I use multifile to load more than one map and check them in hyperspectra a weird checkered visualization happens:
![tempsnip](https://user-images.githubusercontent.com/19801152/125941556-27efca6b-1294-4524-99b7-fe5af3f6060d.png)
and it reads only the image of the first file.

thanks for any help

Gio

**Comment by markotoplak** [comment]
Hi Gio, are you certain only the first image is read? Could you perhaps check the output in the `Data Table` if the images are there? So that we knew whether file reading or display is problematic.

The checkered visualization happens because HyperSpectra could not align all images to the same grid.

Could you share files where this happens? Thanks.

**Comment by GioB81** [comment]
Hi Marko,
with Data Info i get this information:
![image](https://user-images.githubusercontent.com/19801152/126160856-5357904c-6f98-41d7-a103-2561a5d510e3.png)
I got the same behavior with most of the FPA images, at this link you can find some file https://drive.elettra.eu/f/0e2a2220f1f14ab7988e/?dl=1

thanks

Gio

**Comment by markotoplak** [comment]
@GioB81, thanks for the report. Can confirm and will split this issue into two.

We need to redesign how visual images are handled both in Multifile and in Hyperspectra.

Multifile should pass images from all tables through. Before that, we would have to think about the Concatenate widget: how to store the images (and fix Concatenate) so that images from multiple files won't be lost. Multifile should then be solved analogously.

Then, Hyperspectra should allow the display of multiple images. This is probably the easy part: we could have a selector that allows multiple selections.

My first idea: linking images to rows instead of to the table. Then, Concatenate could work sensibly - row metadata would be there from all the input files. Hyperspectra could then show different images that are available among the rows.

@pisarik, do you perhaps have any ideas?

**Comment by pisarik** [comment]
@markotoplak, am I correctly understand in your answer that 
1) The problem with Concatenate is that it doesn't know anything about spectroscopy and it uses general algorithm to concatenate attributes from tables. It seems that [this code](https://github.com/biolab/orange3/blob/704a8379ee7d6345df39f44c8e6cbd5f6ec9ed0d/Orange/data/table.py#L1075) is responsible for table concatenation and problem with current solution.
2) Each row represents a spectrum with `map_x` and `map_y` positions? Does it mean that each row will contain a pointer to the same visible image as metadata? So in Hyperspectra widget we will need to scan all rows and find unique metadata entries?

If it is correct, then it seems that it should work, but in the same time strange to bind visible image to each pixel of spectral image. And number of pixels can be around 100'000, so the scan in python can be slow.

What I like about current situation is that visible images contain all coordiantes and therefore they are completely indepent from spectral images and work on their own (i.e. even in the code there is no any connection between spectral image and visible image). I think we need just add some weak reference to visible image so that user could identify to which spectral image it related and we can continue to keep them independent. For example, we can just save filepath of spectral image into visible image object to provide that reference.

I don't know what is the best, but at least here it is some other options that continue to use table's attributes:

### Workaround 1 (adapt to concatenation as it is)
Now, we are adding attribute called `'visible_images'`, then this attribute just get overwritten during concatenation process. To avoid it we can add attribute called `'visible_images_of_{filepath}'`, then later in Hyperspectra widget just scan all keys that have perfix `visible_images_of_`.

### Workaround 2 (evolve concatenation algorithm)
Develop some general mechanism for concatenation of table attributes that allows to specify which attributes should be overwritten (default) and which attributes should be somehow concatenated as well. I.e. it can take decision based on value of attribute. The value can contain some special field in dict object, or it may have special class that will be treated in other way than other values. Or maybe use some special dict with overriden update method.

### Regarding GUI
In general we have multiple spectral images and each spectral image (SI) can have multiple visible images (VI).
I assume that use cases when SI can have more than 1 VI is very rare.

1) When I select "Show visible images", then over each SI should appear a VI (if possible).
So far we don't need to have a combobox to select images if we assume that each SI can have 0 or 1 VI.
2) To fullfill multiple VI case, I would add a button "Select images", that would open a small window with a single-select-combobox for each SI, rather than put all VI into one multi-select-combobox.

**Comment by stuart-cls** [comment]
I agree with what @pisarik writes. My workaround preference would be to keep `visible_images` dict and then tag members somehow.

One comment about the GUI: I would not assume each spectral image has only 0-1 visible images. I've seen many opus files that have multiple, and the agilent reader supplies 2 visible images for each mosaic by default.

**Comment by pisarik** [comment]
@stuart-cls btw, I just encountered a single opus file that contains 30 small images inside, so I need to scroll through them to visualize a bigger image patch by patch:) 

![image](https://user-images.githubusercontent.com/3684845/223676252-5946063c-258c-4481-b047-c7e0edd211fd.png)

Is it correct to assume that a single opus file always contains images of the same scene? If it is so, then we perhaps could concatenate all the images from a single opus file during loading.

**Comment by stuart-cls** [comment]
@pisarik The acquisition wizard in OPUS is setup in such a way that it is likely they are from the same sample. But it's not enforced, and they could be overlapping / different samples / anything the user does. So it would not be safe to concatenate the images into a single mosaic by default.

That said, it would be reasonable to at least give an option to show all.

Looking back over this issue, for future reference #615 contains some code for merging visible images from multiple files
